src/world/Actor/Player.cpp

```diff
@@ -623,7 +622,28 @@ void Sapphire::Entity::Player::discover( int16_t map_id, int16_t sub_id )
 
   gainExp( exp );
 
+  // gain 10x additional EXP if entire map is completed
+  uint32_t mask = info->discoveryFlag;
+  uint32_t discoveredAreas;
+  if( info->discoveryArrayByte )
+  {
+    discoveredAreas = ( m_discovery[ offset + 1 ] << 8 ) |
+                        m_discovery[ offset ];
+  }
+  else
+  {
+    discoveredAreas = ( m_discovery[ offset + 3 ] << 24 ) |
+                      ( m_discovery[ offset + 2 ] << 16 ) |
+                      ( m_discovery[ offset + 1 ] << 8  ) |
+                        m_discovery[ offset ];
+  }
+
+  bool allDiscovered = ( ( discoveredAreas & mask ) == mask );
 
+  if( allDiscovered )
+  {
+    gainExp( exp * 10 );
```

Comment:
shouldn't it only give 10x exp and not 11x exp? you call gainexp above and then give it again if they clear everything

Reply:
Actually, no, 11x EXP is consistent with retail. When you complete a map on retail, it gives you the regular bonus for completing that portion of the map, then another 10x EXP bonus for completing everything, giving you 11x in total.

That said, this is still a fairly simplistic implementation in other ways. For example, we don't check to make sure that we should even be getting EXP - you don't get EXP for getting discover points in dungeons in retail, for example, but this implementation would give you EXP no matter what. (You may even potentially get the 10x EXP bonus two or three times in a single dungeon for completing each floor map - I'm not sure, I haven't actually tested this scenario as I'm not altogether sure how to detect that yet, and the previous implementation didn't do anything about it either.)

Reply:
there's not really a 'proper' way of checking what kind of an instance something is currently, but I wouldn't worry about it too much anyway because instanced content will need to be looked at again eventually